docs/environment.rst
.. _environment:

=====================
Environment Variables
=====================

ElectrumX takes no command line arguments, instead its behaviour is controlled by
environment variables.  Only a few are required to be given, the rest will have sensible
defaults if not specified.  Many of the defaults around resource usage are conservative; I
encourage you to review them.

.. note:: set :envvar:`SERVICES` appropriately to be able to connect to your server.  For
  clients across the internet to know what services you offer you must advertize your
  services with :envvar:`REPORT_SERVICES`.


Required
========

These environment variables are always required:

.. envvar:: COIN

  Must be a :attr:`NAME` from one of the :class:`Coin` classes in
  `lib/coins.py`_. For example, ``Bitcoin``.

.. envvar:: DB_DIRECTORY

  The path to the database directory.  Relative paths should be
  relative to the parent process working directory.  This is the
  directory of the `run` script if you use it.

.. envvar:: DB_ENGINE

  Database engine for the UTXO and history database.
  Choose one of ``leveldb`` or ``rocksdb``.
  You will need to install the appropriate python package for your engine.
  In ElectrumX 1.x versions, the default was leveldb.
  Warning: It is not possible to switch back and forth,
  the on-disk DB formats are not compatible with each other: you have to resync from genesis.
  LevelDB was written with HDDs in mind. RocksDB is more modern
  and on an SSD takes around ~25% less time than LevelDB to sync from genesis.

.. envvar:: DAEMON_URL

  A comma-separated list of daemon URLs.  If more than one is provided
  ElectrumX will initially connect to the first, and failover to
  subsequent ones round-robin style if one stops working.

  The generic form of a daemon URL is::

     http://username:password@hostname:port/

  The leading ``http://`` is optional, as is the trailing slash.  The
  ``:port`` part is also optional and will default to the standard RPC
  port for :envvar:`COIN` and :envvar:`NET` if omitted.


.. note:: With the above set your server will run and index the chain.  To enable incoming
   connections you must set :envvar:`SERVICES`, and for others to be aware of your server
   set :envvar:`REPORT_SERVICES`.


For the ``run`` script
======================

The following are required if you use the ``run`` script:

.. envvar:: ELECTRUMX

  The path to the electrumx_server script.  Relative paths should
  be relative to the directory of the ``run`` script.

.. envvar:: USERNAME

  The username the server will run as.


Services
========

These two environment variables are comma-separated lists of individual *services*.

A **service** has the general form::

  protocol://host:port

*protocol* is case-insensitive.  The recognised protocols are::

   tcp    Plaintext TCP sockets
   ssl    SSL-encrypted TCP sockets
   ws     Plaintext websockets
   wss    SSL-encrypted websockets
   rpc    Plaintext RPC

In a services list, a protocol can be specified multiple times, with different hosts or
ports.  This might be useful for multi-homed hosts, or if you offer both Tor and clearnet
services.

*host* can be a hostname, an IPv4 address, or an IPv6 address enclosed in square brackets.

*port* is an integer from :const:`1` to :const:`65535` inclusive.

Where documented, one or more of *protocol*, *host* and *port* can be omitted, in which
case a default value will be assumed.

Here are some examples of valid services::

  tcp://host.domain.tld:50001           # Hostname, lowercase protocol, port
  SSL://23.45.67.78:50002               # An IPv4 address, upper-case protocol, port
  rpC://localhost                       # Host as a string, mixed-case protocol, default port
  ws://[1234:5678:abcd::5601]:8000      # Host as an IPv6 address
  wss://h3ubaasdlkheryasd.onion:50001   # Host as a Tor ".onion" address
  rpc://:8000                           # Default host, port given
  host.domain.tld:5151                  # Default protocol, hostname, port
  rpc://                                # RPC protocol, default host and port

.. note:: ElectrumX will not serve any incoming connections until it has fully caught up
          with your bitcoin daemon.  The only exception is local **RPC** connections,
          which are served at any time after the server has initialized.

.. envvar:: SERVICES

  A comma-separated list of services ElectrumX will accept incoming connections for.

  This environment variable determines what interfaces and ports the server listens on, so
  must be set correctly for any connection to the server to succeed.  If unset or empty,
  ElectrumX will not listen for any incoming connections.

  *protocol* can be any recognised protocol.

  *host* defaults to all of the machine's interfaces, except if the protocol is **rpc**,
  when it defaults to :const:`localhost`.

  *port* can only be defaulted for **rpc** where the default is :const:`8000`.

  On most Unix systems ports below 1024 require elevated privileges so choosing a higher
  port is advisable.  On Debian for example, this can be achieved by installing
  libcap2-bin package::

    sudo apt-get update && sudo apt-get -y install libcap2-bin
    sudo setcap cap_net_bind_service=+ep /path/to/electrumx_server

  If any listed service has protocol **ssl** or **wss** then :envvar:`SSL_CERTFILE` and
  :envvar:`SSL_KEYFILE` must be defined.

  Tor **onion** addresses are invalid in :envvar:`SERVICES`.

  Here is an example value of the :envvar:`SERVICES` environment variable::

    tcp://:50001,ssl://:50002,wss://:50004,rpc://

  This serves **tcp**, **ssl**, **wss** on all interfaces on ports 50001, 50002 and 50004
  respectively.  **rpc** is served on its default host :const:`localhost` and default port
  :const:`8000`.

.. envvar:: REPORT_SERVICES

  A comma-separated list of services ElectrumX will advertize and other servers in the
  server network (if peer discovery is enabled), and any successful connection.

  This environment variable must be set correctly, taking account of your network,
  firewall and router setup, for clients and other servers to see how to connect to your
  server.  If not set or empty, no services are advertized.

  The **rpc** protocol, special IP addresses (including private ones if peer discovery is
  enabled), and :const:`localhost` are invalid in :envvar:`REPORT_SERVICES`.

  Here is an example value of the :envvar:`REPORT_SERVICES` environment variable::

    tcp://example.com:50001,ssl://example.com:50002,wss://example.com:50004

  This advertizes **tcp**, **ssl**, **wss** services at :const:`example.com` on ports
  50001, 50002 and 50004 respectively.

.. note:: Certificate Authority-signed certificates don't work over Tor, so you should
          only have Tor services` in :envvar:`REPORT_SERVICES` if yours is self-signed.

.. envvar:: SSL_CERTFILE

  The filesystem path to your SSL certificate file.

  :ref:`SSL certificates`

.. envvar:: SSL_KEYFILE

  The filesystem path to your SSL key file.

  :ref:`SSL certificates`


Miscellaneous
=============

These environment variables are optional:

.. envvar:: LOG_FORMAT

  The Python logging `format string
  <https://docs.python.org/3/library/logging.html#logrecord-attributes>`_
  to use.  Defaults to ``%(levelname)s:%(name)s:%(message)s``.

.. envvar:: LOG_LEVEL

  The default Python logging level, a case-insensitive string.  Useful values
  are 'debug', 'info', 'warning' and 'error'.

.. envvar:: ALLOW_ROOT

  Set this environment variable to anything non-empty to allow running
  ElectrumX as root.

.. envvar:: NET

  Must be a :envvar:`NET` from one of the **Coin** classes in
  `lib/coins.py`_.  Defaults to ``mainnet``.

.. envvar:: DONATION_ADDRESS

  The server donation address reported to Electrum clients.  Defaults
  to empty, which Electrum interprets as meaning there is none.

.. envvar:: BANNER_FILE

  The path to a banner file to serve to clients in Electrum's
  "Console" tab.  Relative file paths must be relative to
  :envvar:`DB_DIRECTORY`.  The banner file is re-read for each new
  client.

  You can place several meta-variables in your banner file, which will be
  replaced before serving to a client.

  + ``$SERVER_VERSION`` is replaced with the ElectrumX version you are
    running, such as ``1.0.10``.
  + ``$SERVER_SUBVERSION`` is replaced with the ElectrumX user agent
    string.  For example, ``ElectrumX 1.0.10``.
  + ``$DAEMON_VERSION`` is replaced with the daemon's version as a
    dot-separated string. For example ``0.12.1``.
  + ``$DAEMON_SUBVERSION`` is replaced with the daemon's user agent
    string.  For example, ``/BitcoinUnlimited:0.12.1(EB16; AD4)/``.
  + ``$DONATION_ADDRESS`` is replaced with the address from the
    :envvar:`DONATION_ADDRESS` environment variable.

  See `here <https://github.com/shsmith/electrumx-banner-updater>`_
  for a script that updates a banner file periodically with useful
  statistics about fees, last block time and height, etc.

.. envvar:: TOR_BANNER_FILE

  As for :envvar:`BANNER_FILE` (which is also the default) but shown
  to incoming connections believed to be to your Tor hidden service.

.. envvar:: ANON_LOGS

  Set to anything non-empty to replace IP addresses in logs with
  redacted text like ``xx.xx.xx.xx:xxx``.  By default IP addresses
  will be written to logs.

.. envvar:: LOG_SESSIONS

  The number of seconds between printing session statistics to the
  log.  The output is identical to the :ref:`sessions` RPC command
  except that :envvar:`ANON_LOGS` is honoured.  Defaults to 3600.  Set
  to zero to suppress this logging.

.. envvar:: REORG_LIMIT

  The maximum number of blocks to be able to handle in a chain
  reorganisation.  ElectrumX retains some fairly compact undo
  information for this many blocks in levelDB.  The default is a
  function of :envvar:`COIN` and :envvar:`NET`; for Bitcoin mainnet it
  is 200.

.. envvar:: EVENT_LOOP_POLICY

  The name of an event loop policy to replace the default asyncio
  policy, if any.  At present only ``uvloop`` is accepted, in which
  case you must have installed the `uvloop`_ Python package.

  If you are not sure what this means leave it unset.

.. envvar:: DROP_CLIENT

  Set a regular expression to disconnect any client based on their
  version string. For example to drop versions from 1.0 to 1.2 use
  the regex ``1\.[0-2]\.\d+``.


Resource Usage Limits
=====================

The following environment variables are all optional and help to limit
server resource consumption and prevent simple DoS.

Address subscriptions in ElectrumX are very cheap - they consume about
160 bytes of memory each and are processed efficiently.  I feel the
two subscription-related defaults below are low and encourage you to
raise them.

.. envvar:: MAX_SESSIONS

  The maximum number of incoming connections.  Once reached, TCP and
  SSL listening sockets are closed until the session count drops
  naturally to 95% of the limit.  Defaults to 1,000.

.. envvar:: MAX_RECV

  The maximum size of an incoming message in bytes, the default is 1,000,000 bytes.
  Note that the smallest sane/safe value for Bitcoin is ~800,100 bytes,
  as the largest standard tx can have a weight of 400K but the protocol hex-encodes that,
  plus there is a few bytes of protocol overhead. Setting this to lower than that
  would preclude clients from broadcasting txs that could propagate over the network.

.. envvar:: MAX_SEND

  The maximum size of a response message to send over the wire, in
  bytes.  Defaults to 8,100,000 (except for AuxPoW coins, which default
  to 10,000,000).  Values smaller than 350,000 are taken as 350,000
  because standard Electrum protocol header "chunk" requests are almost
  that large.

  The Electrum protocol has a flaw in that address histories must be
  served all at once or not at all, an obvious avenue for abuse.
  :envvar:`MAX_SEND` is a stop-gap until the protocol is improved to
  admit incremental history requests.  Each history entry is
  approximately 100 bytes, so the default is equivalent to a history
  limit of around 80,000 entries, which should be ample for most
  legitimate users.  If you use a higher default, bear in mind one
  client can request history for multiple addresses.

  Also note that the largest raw transaction you will be able to serve to a client is
  just under half of :envvar:`MAX_SEND`, as each raw byte becomes 2 hexadecimal
  ASCII characters on the wire.  For Bitcoin, the consensus limit for tx size is
  4 M weight units, while the standardness (policy) limit is 400 K weight units.
  When serialized as bytes, 4 M weight units is 4 MB, and when encoded as hex that is 8 MB.
  The default was chosen so that such a tx could still be served.
  (Note: in multi-party protocol, such as Lightning, a counterparty might collude with a miner
  to get a tx mined. Being able to obtain such a raw tx might be part of the security
  requirements of a protocol.)

.. envvar:: COST_SOFT_LIMIT
.. envvar:: COST_HARD_LIMIT
.. envvar:: REQUEST_SLEEP
.. envvar:: INITIAL_CONCURRENT
.. envvar:: SESSION_GROUP_BY_SUBNET_IPV4
.. envvar:: SESSION_GROUP_BY_SUBNET_IPV6

  All values are integers. :envvar:`COST_SOFT_LIMIT` defaults to :const:`1,000`,
  :envvar:`COST_HARD_LIMIT` to :const:`10,000`, :envvar:`REQUEST_SLEEP` to :const:`2,500`
  milliseconds, and :envvar:`INITIAL_CONCURRENT` to :const:`10` concurrent requests.

  The server prices each request made to it based upon an estimate of the resources needed
  to process it.  Factors include whether the request uses bitcoind, how much bandwidth
  it uses, and how hard it hits the databases.

  To set a base for the units, a :func:`blockchain.scripthash.subscribe` subscription to
  an address with a history of 2 or fewer transactions is costed at :const:`1.0` before
  considering the bandwidth consumed.  :func:`server.ping` is costed at :const:`0.1`.

  As the total cost of a session goes over the soft limit, its requests start to be
  throttled in two ways.  First, the number of requests for that session that the server
  will process concurrently is reduced.  Second, each request starts to sleep a little
  before being handled.

  Before throttling starts, the server will process up to :envvar:`INITIAL_CONCURRENT`
  requests concurrently without sleeping.  As the session cost ranges from
  :envvar:`COST_SOFT_LIMIT` to :envvar:`COST_HARD_LIMIT`, concurrency drops linearly to
  zero and each request's sleep time increases linearly up to :envvar:`REQUEST_SLEEP`
  milliseconds.  Once the hard limit is reached, the session is disconnected.

  In order that non-abusive sessions can continue to be served, a session's cost gradually
  decays over time.  Subscriptions have an ongoing servicing cost, so the decay is slower
  as the number of subscriptions increases.

  If a session disconnects, ElectrumX continues to associate its cost with its IP address,
  so if it immediately reconnects it will re-acquire its previous cost allocation.
  Moreover, sessions are also grouped together based on their IP address subnets, and cost
  is accrued for the whole group. What subnet sizes to use can be configured via
  :envvar:`SESSION_GROUP_BY_SUBNET_IPV4` (by default /24) and
  :envvar:`SESSION_GROUP_BY_SUBNET_IPV6` (by default /48).

  A server operator should experiment with different values according to server loads.  It
  is not necessarily true that e.g. having a low soft limit, decreasing concurrency and
  increasing sleep will help handling heavy loads, as it will also increase the backlog of
  requests the server has to manage in memory.  It will also give a much worse experience
  for genuine connections.

.. envvar:: BANDWIDTH_UNIT_COST

  The number of bytes, sent and received, by a session that is deemed to cost :const:`1.0`.

  The default value :const:`5,000` bytes, meaning the bandwidth cost assigned to a response
  of 100KB is 20.  If your bandwidth is cheap you should probably raise this.

.. envvar:: REQUEST_TIMEOUT

  An integer number of seconds defaulting to :const:`30`.  If a request takes longer than
  this to respond to, either because of request limiting or because the request is
  expensive, the server rejects it and returns a timeout error to the client indicating
  that the server is busy.

  This can help prevent large backlogs of unprocessed requests building up under heavy load.

.. envvar:: SESSION_TIMEOUT

  An integer number of seconds defaulting to :const:`600`.  Sessions that have not sent a
  request for longer than this are disconnected.  Properly functioning clients should send
  a :func:`server.ping` request once roughly 450 seconds have passed since the previous
  request, in order to avoid disconnection.


Peer Discovery
==============

In response to the :func:`server.peers.subscribe` RPC call, ElectrumX
will only return peer servers that it has recently connected to and
verified basic functionality.

If you are not running a Tor proxy ElectrumX will be unable to connect
to onion server peers, in which case rather than returning no onion
peers it will fall back to a hard-coded list.

To give incoming clients a full range of onion servers, you will need
to be running a Tor proxy for ElectrumX to use.

ElectrumX will perform peer-discovery by default and announce itself
to other peers.  If your server is private you may wish to disable
some of this.

.. envvar:: PEER_DISCOVERY

  This environment variable is case-insensitive and defaults to
  ``on``.

  If ``on``, ElectrumX will occasionally connect to and verify its
  network of peer servers.

  If ``off``, peer discovery is disabled and a hard-coded default list
  of servers will be read in and served.  If set to ``self`` then peer
  discovery is disabled and the server will only return itself in the
  peers list.

.. envvar:: PEER_ANNOUNCE

  Set this environment variable to empty to disable announcing itself.
  If not defined, or non-empty, ElectrumX will announce itself to
  peers.

  If peer discovery is disabled this environment variable has no
  effect, because ElectrumX only announces itself to peers when doing
  peer discovery if it notices it is not present in the peer's
  returned list.

.. envvar:: FORCE_PROXY

  By default peer discovery happens over the clear internet.  Set this
  to non-empty to force peer discovery to be done via the proxy.  This
  might be useful if you are running a Tor service exclusively and
  wish to keep your IP address private.

.. envvar:: TOR_PROXY_HOST

  The host where your Tor proxy is running.  Defaults to
  ``localhost``.

  If you are not running a Tor proxy just leave this environment
  variable undefined.

.. envvar:: TOR_PROXY_PORT

  The port on which the Tor proxy is running.  If not set, ElectrumX
  will autodetect any proxy running on the usual ports 9050 (Tor),
  9150 (Tor browser bundle) and 1080 (socks).

.. envvar:: BLACKLIST_URL

  URL to retrieve a list of blacklisted peers.  If not set, a coin-
  specific default is used.


Cache
=====

If synchronizing from the Genesis block your performance might change
by tweaking the cache size.  Cache size is only checked roughly every

docs/changelog.rst
===========
 ChangeLog
===========

Version 2.0 (not yet released)
==============================

TODO expand

* Breaking changes to on-disk database. Server operators need to manually delete old DB,
  no migration path. New DB will be rebuilt while rescanning again from genesis.

   - in return, no more manual "history compaction" and unexpected downtimes every year:
     no more "DB::flush_count overflow" (`spesmilo/electrumx#88`_)
   - the size of the new DB is comparable to the old one, however:

* We now require Bitcoin Core to have `txospenderindex=1` (added in Bitcoin Core 31)
  in addition to `txindex=1`.  This is needed to serve the `blockchain.outpoint.subscribe` RPC
  added in Electrum Protocol 1.7.

  For reference, on Bitcoin mainnet around height=950k,

   - the ElectrumX db uses around 122 GiB (using LevelDB, roughly same for both of e-x 1.x and 2.0),
      - note: using RocksDB 9, it is a bit smaller, around 106 GiB
   - `.bitcoin/blocks/` uses 788 GiB,
   - `.bitcoin/chainstate/` uses 12 GiB,
   - `.bitcoin/indexes/txindex/` uses 66 GiB,
   - `.bitcoin/indexes/txospenderindex/` uses 88 GiB (new!)

  - This also means that we now require Bitcoin Core 31.0 or newer (for `COIN=Bitcoin`).

* env: the previously optional envvar `DB_ENGINE`, is now mandatory.

  - choose either `leveldb` or `rocksdb`
  - in ElectrumX 1.x versions, the default was leveldb.

  You need to install the appropriate dependencies for your engine, see the `[leveldb]` and
  `[rocksdb]` pip extras and the "Database Engine" section of the HOWTO.

  Previously RocksDB was difficult to use as the python bindings for it have been unmaintained for years.
  By the super-slop-powers of LLMs, we revived the python bindings as `rocksdb-ng` and made it compatible with modern
  cpython, cython, and rocksdb. The changes are reviewable and not really slop. (see `spesmilo/electrumx#347`_)
  rocksdb-ng (starting with version 2.3) is also compatible with `free-threaded python`_, see "performance" section.

  LevelDB was written with HDDs in mind. RocksDB is more modern
  and on an SSD takes around ~25% less time than LevelDB to sync from genesis.
  Given that this release contains breaking DB changes and a resync is needed anyway, maybe try out RocksDB :)

* performance:

   - the BlockProcessor is now massively multi-threaded, both inside a single block and across a span of blocks.
     This can mainly only be taken advantage of if the GIL (Global Interpreter Lock) is disabled, that is,
     when using a `free-threaded python`_ interpreter: e.g. cpython `3.14t`, compiled with `--disable-gil`.
     On a decent modern machine with a multi-core CPU, this has been measured to cut initial sync time
     from genesis to a bit less than half. OTOH, much more RAM is required for the parallel processing
     (example: approx 2 GB when syncing with GIL, 8 GB when syncing without GIL).

   - processing a single block at the tip, when already caught up, is expected to take around half the time,
     regardless of the GIL. Still, without the GIL, it is even faster.

   - for more details, see `spesmilo/electrumx#369`_

* protocol:
   - new: implement electrum protocol version 1.7  (`spesmilo/electrum-protocol#2`_).
     The min supported protocol version remains 1.4, the max is now 1.7.
* performance: json (de)serialization can now optionally use the Rust-based
  :code:`orjson` library (via the :code:`orjson` pip extra), replacing the
  previous :code:`ujson` and :code:`rapidjson` extras.


Version 1.20.0 (03 June 2026)
=============================

Small fixes and general maintenance.

After this release, the master branch will move towards a 2.0 release, which will require
a resync from genesis. So for a while, master might be less stable than usual.

* fix: lib/tx.py: vsize calculation was off by 6-8 vbytes (`spesmilo/electrumx#328`_).
  Added tests. The discrepancy was externally observable via the `mempool.get_fee_histogram` RPC.
* fix: don't get stuck during shutdown waiting for SessionManager to stop (`spesmilo/electrumx#339`_)
* fix: protocol: name of arg for broadcast_package RPC: tx_package->raw_txs (`c9b72e7e`_)
* changed: peer discovery: allow short chain splits (`spesmilo/electrumx#351`_)
* changed: rm 'attrs' as a direct dependency, use stdlib `dataclasses` instead (`spesmilo/electrumx#345`_)
* new: daemon: check early "txindex" is enabled for bitcoind, error if missing (`spesmilo/electrumx#349`_)
* new: coins: add Bitcoin mutinynet (`spesmilo/electrumx#352`_)
* maintenance: coins: some clean-up (`spesmilo/electrumx#344`_, `spesmilo/electrumx#348`_, ...)
* maintenance: moved CI from Cirrus CI to Github Actions (`spesmilo/electrumx#340`_)

(We also made some changes to make RocksDB practical and nicer to use, but if you currently use LevelDB,
maybe consider waiting for 2.0 before you resync the DB)


Version 1.19.0 (11 Nov 2025)
=============================

* protocol:
   - new: implement electrum protocol version 1.6 (`spesmilo/electrumx#317`_, ...)
     (`spesmilo/electrum-protocol#6`_).
     The min supported protocol version remains 1.4, the max is now 1.6.
* coins:
   - changed: for `COIN=Bitcoin`, there is now a min required bitcoind version:
     Bitcoin Core 28.0 (or Knots 28) (`spesmilo/electrumx#316`_).
     This is due to protocol 1.6 requiring a bitcoind with working `submitpackage` RPC.
* security:
   - fix DOS vector: if MAX_SESSIONS was reached, when using python 3.12+, the server
     stopped accepting new incoming sessions until full-restart (`spesmilo/electrumx#312`_)
* session:
   - changed: add warmup budget to PaddedRSTransport (`spesmilo/electrumx#323`_):
     this is a small relaxation of the traffic analysis countermeasures added in 1.18,
     where the first 1024 bytes we send are now exempt from buffering (and hence delays)
   - fix ReplyAndDisconnect for PaddedRSTransport: flush message buffer (`spesmilo/electrumx#322`_)
* env:

src/electrumx/server/env.py
# Copyright (c) 2016, Neil Booth
#
# All rights reserved.
#
# See the file "LICENCE" for information about the copyright
# and warranty status of this software.

'''Class for handling environment configuration and defaults.'''


import re
from ipaddress import IPv4Address, IPv6Address
from typing import Type, Sequence

from aiorpcx import Service, ServicePart

import electrumx
from electrumx.lib.coins import Coin
from electrumx.lib.env_base import EnvBase


class ServiceError(Exception):
    pass


class Env(EnvBase):
    '''Wraps environment configuration. Optionally, accepts a Coin class
       as first argument to have ElectrumX serve custom coins not part of
       the standard distribution.
    '''

    # Peer discovery
    PD_OFF, PD_SELF, PD_ON = ('OFF', 'SELF', 'ON')
    SSL_PROTOCOLS = {'ssl', 'wss'}
    KNOWN_PROTOCOLS = {'ssl', 'tcp', 'ws', 'wss', 'rpc'}

    coin: Type[Coin]

    def __init__(self, coin=None):
        super().__init__()
        self.obsolete(["MAX_SUBSCRIPTIONS", "MAX_SUBS", "MAX_SESSION_SUBS", "BANDWIDTH_LIMIT",
                       "HOST", "TCP_PORT", "SSL_PORT", "RPC_HOST", "RPC_PORT", "REPORT_HOST",
                       "REPORT_TCP_PORT", "REPORT_SSL_PORT", "REPORT_HOST_TOR",
                       "REPORT_TCP_PORT_TOR", "REPORT_SSL_PORT_TOR"])

        # Core items

        self.db_dir = self.required('DB_DIRECTORY')
        self.db_engine = self.db_engine_enum()
        self.daemon_url = self.required('DAEMON_URL')
        if coin is not None:
            assert issubclass(coin, Coin)
            self.coin = coin
        else:
            coin_name = self.required('COIN').strip()
            network = self.default('NET', 'mainnet').strip()
            self.coin = Coin.lookup_coin_class(coin_name, network)

        # Peer discovery

        self.peer_discovery = self.peer_discovery_enum()
        self.peer_announce = self.boolean('PEER_ANNOUNCE', True)
        self.force_proxy = self.boolean('FORCE_PROXY', False)
        self.tor_proxy_host = self.default('TOR_PROXY_HOST', 'localhost')
        self.tor_proxy_port = self.integer('TOR_PROXY_PORT', None)

        # Misc
        self.banner_file = self.default('BANNER_FILE', None)
        self.tor_banner_file = self.default('TOR_BANNER_FILE',
                                            self.banner_file)
        self.anon_logs = self.boolean('ANON_LOGS', False)
        self.log_sessions = self.integer('LOG_SESSIONS', 3600)
        self.log_level = self.default('LOG_LEVEL', 'info').upper()
        self.donation_address = self.default('DONATION_ADDRESS', '')
        self.drop_client = self.custom("DROP_CLIENT", None, re.compile)
        self.blacklist_url = self.default('BLACKLIST_URL', self.coin.BLACKLIST_URL)
        self.cache_MB = self.integer('CACHE_MB', 1200)
        self.reorg_limit = self.integer('REORG_LIMIT', self.coin.REORG_LIMIT)
        self.daemon_poll_interval_blocks_msec = self.integer('DAEMON_POLL_INTERVAL_BLOCKS', 5000)
        self.daemon_poll_interval_mempool_msec = self.integer('DAEMON_POLL_INTERVAL_MEMPOOL', 5000)

        # Server limits to help prevent DoS

        self.max_send = self.integer('MAX_SEND', self.coin.DEFAULT_MAX_SEND)
        self.max_recv = self.integer('MAX_RECV', self.coin.DEFAULT_MAX_RECV)
        self._try_raising_rlimit_nofile()  # do this before setting max_sessions
        self.max_sessions = self.sane_max_sessions()
        self.cost_soft_limit = self.integer('COST_SOFT_LIMIT', 1000)
        self.cost_hard_limit = self.integer('COST_HARD_LIMIT', 10000)
        self.bw_unit_cost = self.integer('BANDWIDTH_UNIT_COST', 5000)
        self.initial_concurrent = self.integer('INITIAL_CONCURRENT', 10)
        self.request_sleep = self.integer('REQUEST_SLEEP', 2500)
        self.request_timeout = self.integer('REQUEST_TIMEOUT', 30)
        self.session_timeout = self.integer('SESSION_TIMEOUT', 600)
        self.session_group_by_subnet_ipv4 = self.integer('SESSION_GROUP_BY_SUBNET_IPV4', 24)
        self.session_group_by_subnet_ipv6 = self.integer('SESSION_GROUP_BY_SUBNET_IPV6', 48)
        self._check_and_fix_cost_limits()

        # Services last - uses some env vars above

        self.services = self.services_to_run()
        if {service.protocol for service in self.services}.intersection(self.SSL_PROTOCOLS):
            self.ssl_certfile = self.required('SSL_CERTFILE')
            self.ssl_keyfile = self.required('SSL_KEYFILE')
        self.report_services = self.services_to_report()

    def _try_raising_rlimit_nofile(self) -> None:
        """Try raising the max num open file soft limit to the hard limit."""
        try:
            import resource
        except ImportError:
            return  # No resource module on Windows
        soft, hard = resource.getrlimit(resource.RLIMIT_NOFILE)
        if soft >= hard:
            return
        try:
            resource.setrlimit(resource.RLIMIT_NOFILE, (hard, hard))
        except (ValueError, OSError):
            self.logger.warning(
                f"error raising RLIMIT_NOFILE (ulimit -n) soft limit. Stuck with {soft}. "
                f"You must ensure that electrumx has a large open file limit. See HOWTO in docs.")
        else:
            self.logger.info(
                f"successfully raised RLIMIT_NOFILE (ulimit -n) soft limit from {soft} to {hard}")

    def sane_max_sessions(self):
        '''Return the maximum number of sessions to permit.  Normally this
        is MAX_SESSIONS.  However, to prevent open file exhaustion, adjust
        downwards if running with a small open file rlimit.'''
        env_value = self.integer('MAX_SESSIONS', 1000)
        # No resource module on Windows
        try:
            import resource
            nofile_limit = resource.getrlimit(resource.RLIMIT_NOFILE)[0]
            # We give the DB 250 files; allow ElectrumX 100 for itself
            value = max(0, min(env_value, nofile_limit - 350))
            if value < env_value:
                self.logger.warning(
                    f'lowered maximum sessions from {env_value:,d} to '
                    f'{value:,d} because your open file limit is '
                    f'{nofile_limit:,d}'
                )
        except ImportError:
            value = 512  # that is what returned by stdio's _getmaxstdio()
        return value

    def _check_and_fix_cost_limits(self):
        if self.cost_hard_limit < self.cost_soft_limit:
            raise self.Error(f"COST_HARD_LIMIT must be >= COST_SOFT_LIMIT. "
                             f"got (COST_HARD_LIMIT={self.cost_hard_limit} "
                             f"and COST_SOFT_LIMIT={self.cost_soft_limit})")
        # hard limit should be strictly higher than soft limit (unless both are 0)
        if self.cost_hard_limit == self.cost_soft_limit and self.cost_soft_limit > 0:
            self.logger.info("found COST_HARD_LIMIT == COST_SOFT_LIMIT. "
                             "bumping COST_HARD_LIMIT by 1.")
            self.cost_hard_limit = self.cost_soft_limit + 1

    def _parse_services(self, services_str, default_func) -> Sequence[Service]:
        result = []
        for service_str in services_str.split(','):
            if not service_str:
                continue
            try:
                service = Service.from_string(service_str, default_func=default_func)
            except Exception as e:
                raise ServiceError(f'"{service_str}" invalid: {e}') from None
            if service.protocol not in self.KNOWN_PROTOCOLS:
                raise ServiceError(f'"{service_str}" invalid: unknown protocol')
            result.append(service)

        # Find duplicate addresses
        service_map = {service.address: [] for service in result}
        for service in result:
            service_map[service.address].append(service)
        for address, services in service_map.items():
            if len(services) > 1:
                raise ServiceError(f'address {address} has multiple services')

        return result

    def services_to_run(self) -> Sequence[Service]:
        def default_part(protocol, part):
            return default_services.get(protocol, {}).get(part)

        default_services = {protocol: {ServicePart.HOST: 'all_interfaces'}
                            for protocol in self.KNOWN_PROTOCOLS}
        default_services['rpc'] = {ServicePart.HOST: 'localhost', ServicePart.PORT: 8000}
        services = self._parse_services(self.default('SERVICES', ''), default_part)

        # Find onion hosts
        for service in services:
            if str(service.host).endswith('.onion'):
                raise ServiceError(f'bad host for SERVICES: {service}')

        return services

    def services_to_report(self):
        services = self._parse_services(self.default('REPORT_SERVICES', ''), None)

        for service in services:
            if service.protocol == 'rpc':
                raise ServiceError(f'bad protocol for REPORT_SERVICES: {service.protocol}')
            if isinstance(service.host, (IPv4Address, IPv6Address)):
                ip_addr = service.host
                if (ip_addr.is_multicast or ip_addr.is_unspecified or
                        (ip_addr.is_private and self.peer_announce)):
                    raise ServiceError(f'bad IP address for REPORT_SERVICES: {ip_addr}')
            elif service.host.lower() == 'localhost':
                raise ServiceError(f'bad host for REPORT_SERVICES: {service.host}')

        return services

    def peer_discovery_enum(self):
        pd = self.default('PEER_DISCOVERY', 'on').strip().lower()
        if pd in ('off', ''):
            return self.PD_OFF
        elif pd == 'self':
            return self.PD_SELF
        else:
            return self.PD_ON

    def db_engine_enum(self) -> str:
        from electrumx.server.storage import list_db_engine_choices
        choices = list_db_engine_choices()
        db_eng = self.default('DB_ENGINE', None)

        if db_eng is None:
            raise self.Error(
                f"required envvar DB_ENGINE not set.\n"
                f"Choose one from: {choices}.\n"
                f"The corresponding dependencies also need to be installed. See pip extras.\n"
                f"In ElectrumX 1.x versions, the default was leveldb.\n"
                f"Warning: It is not possible to switch back and forth, "
                f"the on-disk DB formats are not compatible with each other: when switching, "
                f"you have to resync from genesis.\n"
                f"LevelDB was written with HDDs in mind. RocksDB is more modern and performs better (on an SSD)."
            )

        db_eng = db_eng.lower()
        if db_eng in choices:
            return db_eng
        else:
            raise self.Error(
                f"bad value for envvar DB_ENGINE: {db_eng!r}. Choose one from: {choices}."
            )
